invokeai/backend/stable_diffusion/diffusion/custom_attention.py

```diff
@@ -0,0 +1,342 @@
+from dataclasses import dataclass
+from typing import List, Optional
+
+import torch
+import torch.nn.functional as F
+from diffusers.models.attention_processor import Attention
+
+import invokeai.backend.util.logging as logger
+from invokeai.app.services.config.config_default import get_config
+from invokeai.backend.ip_adapter.ip_attention_weights import IPAttentionProcessorWeights
+from invokeai.backend.stable_diffusion.diffusion.regional_ip_data import RegionalIPData
+from invokeai.backend.stable_diffusion.diffusion.regional_prompt_data import RegionalPromptData
+from invokeai.backend.util.devices import TorchDevice
+
+
+@dataclass
+class IPAdapterAttentionWeights:
+    ip_adapter_weights: IPAttentionProcessorWeights
+    skip: bool
+
+
+class CustomAttnProcessor:
+    """A custom implementation of attention processor that supports additional Invoke features.
+    This implementation is based on
+    SlicedAttnProcessor (https://github.com/huggingface/diffusers/blob/fcfa270fbd1dc294e2f3a505bae6bcb791d721c3/src/diffusers/models/attention_processor.py#L1616)
+    AttnProcessor2_0 (https://github.com/huggingface/diffusers/blob/fcfa270fbd1dc294e2f3a505bae6bcb791d721c3/src/diffusers/models/attention_processor.py#L1204)
+    Supported custom features:
+    - IP-Adapter
+    - Regional prompt attention
+    """
+
+    def __init__(
+        self,
+        ip_adapter_attention_weights: Optional[List[IPAdapterAttentionWeights]] = None,
+    ):
+        """Initialize a CustomAttnProcessor.
+        Note: Arguments that are the same for all attention layers are passed to __call__(). Arguments that are
+        layer-specific are passed to __init__().
+        Args:
+            ip_adapter_weights: The IP-Adapter attention weights. ip_adapter_weights[i] contains the attention weights
+                for the i'th IP-Adapter.
+        """
+
+        self._ip_adapter_attention_weights = ip_adapter_attention_weights
+        self.attention_type, self.slice_size = self._select_attention()
+
+        # inspect didn't work because it's native function
+        # In 2.0 torch there no scale argument in sdp, it's added in 2.1
+        # Probably can selected based on torch version instead
+        try:
+            F.scaled_dot_product_attention(torch.zeros(1,1), torch.zeros(1,1), torch.zeros(1,1), scale=0.5)
+            self.scaled_sdp = True
+        except:
+            self.scaled_sdp = False
+
+    def _select_attention(self):
+        config = get_config()
+        attention_type = config.attention_type
+        if attention_type in ["normal", "xformers"]:
```

Comment:
This warning and deprecation code is repeated of `diffusers_pipeline.py`. Could be turned into a common function.